Data/HashMap/Internal.hs

```diff
@@ -803,37 +804,38 @@ insert k v m = insert' (hash k) k v m
 {-# INLINABLE insert #-}
 
 insert' :: Eq k => Hash -> k -> v -> HashMap k v -> HashMap k v
-insert' h0 k0 v0 m0 = go h0 k0 v0 0 m0
+insert' h0 k0 v0 m0 =
+    case go h0 k0 v0 0 m0 of
+      NoChange  -> m0
+      Changed m -> m
   where
-    go !h !k x !_ Empty = Leaf h (L k x)
+    go !h !k x !_ Empty = Changed (Leaf h (L k x))
     go h k x s t@(Leaf hy l@(L ky y))
         | hy == h = if ky == k
                     then if x `ptrEq` y
-                         then t
-                         else Leaf h (L k x)
-                    else collision h l (L k x)
-        | otherwise = runST (two s h k x hy t)
-    go h k x s t@(BitmapIndexed b ary)
+                         then NoChange
+                         else Changed (Leaf h (L k x))
+                    else Changed $ collision h l (L k x)
+        | otherwise = Changed $ runST (two s h k x hy t)
```

Comment:
I suspect that we're less lazy in the result of `two` now, and that this is the actual source of the minor performance improvements.

Reply:
I should try to make `two` itself more strict: Force the result of the recursive application of `go`, force the `return`ed `BitmapIndexed` nodes.